File: DSA/src/Questions/BinaryTree/NextRightPointers.java
package Questions.BinaryTree;

public class NextRightPointers {
    public static class Node {
        int val;
        Node left;
        Node right;
        Node next;

        public Node() {
        }

        public Node(int val) {
            this.val = val;
        }

        public Node(int val, Node left, Node right, Node next) {
            this.val = val;
            this.left = left;
            this.right = right;
            this.next = next;
        }
    }
    public Node connect(Node root) {
        // System.out.println("root: " + root.val);
        if (root == null) return null;

        Node currentLevelStart = root; // Start with the root of the tree

        while (currentLevelStart.left != null) { // While there's a next level
            Node current = currentLevelStart;

            while (current != null) { // Traverse the current level
                current.left.next = current.right; // Connect left child to right child

                if (current.next != null) { // Connect right child to next node's left child
                    current.right.next = current.next.left;
                }

                current = current.next; // Move to the next node in the current level
            }

            currentLevelStart = currentLevelStart.left; // Move down to the next level
        }

        System.out.println("root: " + root.val);
        return root; // Return the modified tree
    }
    public static void main(String[] args) {
        NextRightPointers nextRightPointers = new NextRightPointers();
        Node root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        root.right.left = new Node(6);
        root.right.right = new Node(7);
        // Tree
        //      1
        //    /  \
        //   2    3
        //  / \  / \
        // 4   5 6  7
        Node result = nextRightPointers.connect(root);
        // Traverse through the tree and print the next pointers
        while (result != null) {
            Node current = result;
            while (current != null) {
                System.out.print(current.val + " -> ");
                current = current.next;
            }
            System.out.println("null");
            result = result.left;
        }

    }
}

File: DSA/Python/Questions/BinaryTree/NextRightPointers.py
"""
# Question: Populating Next Right Pointers in Each Node
# Link: https://leetcode.com/problems/populating-next-right-pointers-in-each-node/

# Time Complexity: O(N)
# Space Complexity: O(1)

# Algorithm:
# 1. Level by level traversal
# 2. Use next pointers
# 3. Connect nodes at each level
"""


class Node:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None
        self.next = None


class NextRightPointers:
    def connect(self, root: Node) -> Node:
        if not root:
            return None

        leftmost = root
        while leftmost.left:
            head = leftmost
            while head:
                head.left.next = head.right
                if head.next:
                    head.right.next = head.next.left
                head = head.next
            leftmost = leftmost.left

        return root


def main():
    solution = NextRightPointers()
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    solution.connect(root)
    print(root.left.next.val)  # Should print 3


if __name__ == "__main__":
    main()
